### ISSUE TITLE
[Op] In-layer gradient norm problem running on multiple devices

### ISSUE THREAD

**Issue by zhreshold** [apache/incubator-mxnet#6710]
Just realized that the normalization mode 'batch' and 'valid' are problematic in SoftmaxOutput and MakeLoss layers if we dispatch data to multiple devices.

```
'null': do nothing.
'batch': divide the gradient by the batch size.
'valid': divide the gradient by the number of instances which are not ignored.
```

Suppose we have batch size 64, using 'batch' norm,  the gradients we get will be (g1/32 + g2/32) on 2 gpus, and (g1 + g2) / 64 on single gpu, where g1 and g2 are first half/second half of the gradients, respectively.

Things become more complicated in 'valid' norm, since we might get uneven distribution for valid samples, thus the gradients will become very different.

Summed up, we got different results using --gpus 0/--gpus 0,1 with exactly same hyperparameter set as long as we use 'batch' or 'valid' norm in these layers, which is not a desired behavior in my mind.

Any idea? @piiswrong @mli

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
